### DISCUSSION TITLE
BMP280 and AHT20

### DISCUSSION THREAD

**Discussion by flexiti** [HASwitchPlate/openHASP#666] [Show and tell]
In my board (ESP32-4848S040)  connected in parallel to G911, IO19(SDA)  and IO45(SCL)

https://pl.aliexpress.com/item/1005003123605458.html?spm=a2g0o.productlist.main.25.17be6b35hjl9XJ&algo_pvid=39648458-5656-402a-85c0-6404d459d7ec&algo_exp_id=39648458-5656-402a-85c0-6404d459d7ec-12&pdp_npi=4%40dis%21PLN%213.63%213.63%21%21%210.90%210.90%21%40211b613117100044365095668ef3f6%2112000024222833412%21sea%21PL%21121852947%21&curPageLogUid=MiLzWAR3m5dl&utparam-url=scene%3Asearch%7Cquery_from%3A

my_custom.cpp

```
#include <Adafruit_BMP280.h>
#include <Adafruit_AHTX0.h>
Adafruit_BMP280 bmp; // I2C
Adafruit_AHTX0 aht;

float BmpTemperature;
float BmpPreasure; 
float AthTemperature;
float AthHumanidti; 

void custom_setup()
{
bmp.begin(0x77);  //0x77
aht.begin();            //0x38
}

void custom_every_5seconds()
{

 BmpTemperature = bmp.readTemperature();
    BmpPreasure = bmp.readPressure();

    Serial.print(F("#### Temperature = "));
    Serial.print(BmpTemperature);
    Serial.println(" *C");

    Serial.print(F("#### Pressure = "));
    Serial.print(BmpPreasure/100);
    Serial.println(" Pa");

    
            page=1;
            id = 10;   //// label on page 1
            lv_obj_t* MyTemp10 = hasp_find_obj_from_page_id(page, id);
            if(MyTemp10) //return; // object doesn't exist
                    lv_label_set_text_fmt(MyTemp10,"#FFFFFF %0.1f°C",BmpTemperature);
            page=1;
            id = 11;    //// label on page 1
            lv_obj_t* MyTemp11 = hasp_find_obj_from_page_id(page, id);
            if(MyTemp11) //return; // object doesn't exist
                    lv_label_set_text_fmt(MyTemp11,"#FFFFFF %0.0f Hp",BmpPreasure /100);
   sensors_event_t humidity, temp;
      aht.getEvent(&humidity, &temp);// populate temp and humidity objects with fresh data
      AthTemperature = temp.temperature;
      AthHumanidti = humidity.relative_humidity;
 
            page=1;
            id = 12;   // label on page 1
            lv_obj_t* MyTemp12 = hasp_find_obj_from_page_id(page, id);
            if(MyTemp12) //return; // object doesn't exist
                    lv_label_set_text_fmt(MyTemp12,"#FFFFFF %0.1f°C",AthTemperature);
            page=1;
            id = 13;  // label on page 1
            lv_obj_t* MyTemp13 = hasp_find_obj_from_page_id(page, id);
            if(MyTemp13) //return; // object doesn't exist
                    lv_label_set_text_fmt(MyTemp13,"#FFFFFF %0.1f%s",AthHumanidti,"%");
}

```
https://github.com/adafruit/Adafruit_AHTX0/tree/master
https://registry.platformio.org/libraries/adafruit/Adafruit%20BMP280%20Library

platformio.ini - add line:
 adafruit/Adafruit AHTX0@^2.0.5
adafruit/Adafruit BMP280 Library@^2.6.8

😏
![ESP32_view1](https://github.com/HASwitchPlate/openHASP/assets/11538048/62ff0092-c8f8-46cb-ab11-091871797a53)

`

**Comment by greeaaby** [comment]
Problem with compiling - can you pleas help? 
i pasted 
adafruit/Adafruit AHTX0@^2.0.5
adafruit/Adafruit BMP280 Library@^2.6.8
in platformio.ini , but get this error:
**_gitpod /workspace/openHASP $ platformio run
InvalidProjectConfError: Invalid '/workspace/openHASP/platformio.ini' (project configuration file): 'File contains no section headers.
file: '/workspace/openHASP/platformio.ini', line: 9
'adafruit/Adafruit AHTX0@^2.0.5\n''_**

**Comment by flexiti** [comment]
@greeaaby  
Did you put it in the right place?
Check (delete?) \n -  `adafruit/Adafruit AHTX0@^2.0.5` **\n** in your code

 ```
   -D HASP_VER_REV=0-rc11
    ;-D HASP_VER_REV=4-rc1
    ${override.build_flags}

; -- Shared library dependencies in all environments
; Warning : don't put comments after github links => causes infinite download loop
lib_deps =
    robtillaart/DS18B20@^0.2.2 
    adafruit/Adafruit BMP280 Library@^2.6.8
    enjoyneering/AHT10@^1.1.0
    adafruit/Adafruit AHTX0@^2.0.5
    bblanchon/ArduinoJson@^6.21.5
    ;git+https://github.com/fvanroie/ConsoleInput.git
    ;git+https://github.com/andrethomas/TasmotaSlave.git
    ;git+https://github.com/lvgl/lvgl.git
    ;git+https://github.com/lvgl/lvgl.git#release/v7
    git+https://github.com/HASwitchPlate/lvgl.git#release/v7
    ;git+https://github.com/lvgl/lv_fs_if.git#release/v7
    git+https://github.com/lvgl/lv_lib_png.git#release/v7
    git+https://github.com/lvgl/lv_lib_bmp.git#release/v7
    git+https://github.com/lvgl/lv_lib_gif.git#release/v7
```

**Reply by greeaaby** [reply]
Ok, 1 step .... Thank you a lot ...
now i'm struggeling with the custom.h are put in src/custom
Both custom.h and custom.cpp 
Errors i get:
/* MIT License - Copyright (c) 2019-2024 Francis Van Roie
   For full license information read the LICENSE file in the project folder */

// USAGE: - Copy this file and rename it to my_custom.h
//        - uncomment in your user_config_override.h the line containing #define HASP_USE_CUSTOM 1
//

#ifndef HASP_CUSTOM_H
#define HASP_CUSTOM_H

#include "hasplib.h"
#if defined(HASP_USE_CUSTOM)

/* This function is called at boot */
void custom_setup();

/* This function is called every itteration of the main loop */
void custom_loop();

/* This function is called every second */
void custom_every_second();

/* This function is called every 5 seconds */
void custom_every_5seconds();

/* return true if the pin used by the custom code */
bool custom_pin_in_use(uint8_t pin);

/* Add a key which defines a JsonObject to add to the sensor JSON output */
void custom_get_sensors(JsonDocument& doc);

/* Receive custom topic & payload messages */
void custom_topic_payload(const char* topic, const char* payload, uint8_t source);

#endif // HASP_USE_CUSTOM

#endif // HASP_CUSTOM_H
............................. 
The  custom.h i used:
src/my_custom.h

#ifndef HASP_CUSTOM_H
#define HASP_CUSTOM_H
#if defined(HASP_USE_CUSTOM)

#include "hasplib.h"

void custom_setup();
void custom_loop();
void custom_every_second();
void custom_every_5seconds();
bool custom_pin_in_use(uint8_t pin);
void custom_get_sensors(JsonDocument& doc);
void custom_topic_payload(const char* topic, const char* payload, uint8_t source);

#endif // HASP_USE_CUSTOM
#endif // HASP_CUSTOM_H
.................................

**Comment by flexiti** [comment]
In user_config_override.h make change as below, and then make file: my_custom.cpp and my_custom.h (or another name)
and change this line in this file to:

```
my_custom.cpp
#if defined(HASP_USE_CUSTOM) && true // <-- set this to true in your code
```


```
user_config_override.h
/***************************************************
 *        Other Settings
 **************************************************/
//#define HASP_USE_MDNS 0                             // Disable MDNS
#define HASP_USE_CUSTOM 1                          // Enable compilation of custom code from /src/custom
//#define HASP_USE_HA                                 // Enable Home Assistant auto-discovery
```

**Reply by XiaoTuXing** [reply]
can you help me

**Reply by flexiti** [reply]
At first remove print from custom loop, it will print all the time and block all

**Comment by greeaaby** [comment]
Thank you - Now i made another big step.
in custom.cpp under "void custom_every_5seconds()"
i pasted:
       uint8_t page = 1;   // the page of the object you want to rotate
       uint8_t id   = 150; // the id of the object you want to rotate
Now i have a display!!

**Comment by XiaoTuXing** [comment]
**openHASP\src\my_custom.cpp Why not work，HASP_USE_CUSTOM  set to 1**
#include "hasplib.h"

#if defined(HASP_USE_CUSTOM) && true // <-- set this to true in your code

#include "hasp_debug.h"

void custom_setup()
{
    // Initialization code here
    randomSeed(millis());
}

void custom_loop()
{
    // Non-blocking code here, this should execute very fast!
    Serial.print("temp_custom_loop");
}

void custom_every_second()
{
    Serial.print("#");
}

**Reply by XiaoTuXing** [reply]
you are right，but it print no any word about this file；
i find USE_CONFIG_OVERRIDE is not definition，so I modified openHASP\src\hasplib.h 
`
#ifdef USE_CONFIG_OVERRIDE
#include "user_config_override.h"
#endif` 
TO
`#include "user_config_override.h"
#ifdef USE_CONFIG_OVERRIDE

#endif`

**Reply by flexiti** [reply]
in user_config_override:

//#define HASP_USE_MDNS 0                             // Disable MDNS
#**define HASP_USE_CUSTOM 1**                           // Enable compilation of custom code from /src/custom
//#define HASP_USE_HA                                 // Enable Home Assistant auto-discovery
//#define HASP_START_CONSOLE 0                        // Disable starting of serial console at boot
//#define HASP_START_TELNET 0                         // Disable starting of telnet service at boot
//#define HASP_START_HTTP 0                           // Disable starting of web interface at boot
//#define HASP_START_FTP 0                            // Disable starting of ftp server at boot
//#define LV_MEM_SIZE (64 * 1024U)                    // 64KiB of lvgl memory (default 48)
//#define LV_VDB_SIZE (32 * 1024U)                    // 32KiB of lvgl draw buffer (default 32)
//#define HASP_DEBUG_OBJ_TREE                         // Output all objects to the log on page changes
#define HASP_LOG_LEVEL LOG_LEVEL_DEBUG //LOG_LEVEL_DEBUG            // LOG_LEVEL_* can be DEBUG, VERBOSE, TRACE, INFO, WARNING, ERROR, CRITICAL, ALERT, FATAL, SILENT
//#define HASP_LOG_TASKS                              // Also log the Taskname and watermark of ESP32 tasks

**Reply by XiaoTuXing** [reply]
i know,i think user_config_override.h need to include ,USE_CONFIG_OVERRIDE ability can be read

**Comment by flexiti** [comment]
void custom_loop()
{
// Non-blocking code here, this should execute very fast!
//////////////////Serial.print("temp_custom_loop");    remove !!!
}

**Comment by alarik84** [comment]
Hi.
I followed this example and was able to run AHT21. The only unexpected thing is it reports temperature in negative.
Any idea why and how to solve?
